scripts/lib/ts-generator.js

```diff
@@ -0,0 +1,437 @@
+        code += tabs(tabCount) + 'export type ' + shapeKey + ' = Date;\n';
+    } else if (type === 'boolean') {
+        code += tabs(tabCount) + 'export type ' + shapeKey + ' = boolean;\n';
+    } else if (type === 'blob' || type === 'binary') {
+        code += tabs(tabCount) + 'export type ' + shapeKey + ' = Buffer|Uint8Array|Blob|string;\n';
+    }
+    return code;
+};
+
+/**
+ * Generates a class method type for an operation.
+ */
+TSGenerator.prototype.generateTypingsFromOperations = function generateTypingsFromOperations(className, operation, tabCount) {
+    var code = '';
+    tabCount = tabCount || 0;
+    var tabs = this.tabs;
+
+    var input = operation.input;
+    var output = operation.output;
+    var operationName = operation.name.charAt(0).toLowerCase() + operation.name.substring(1);
+
+    var inputShape = input ? className + '.' + input.shape : '{}';
+    var outputShape = output ? className + '.' + output.shape : '{}';
+
+    code += this.generateDocString(operation.documentation, tabCount);
+    code += tabs(tabCount) + operationName + '(params: ' + inputShape + ', callback?: (err: AWSError, data: ' + outputShape + ') => void): Request<' + outputShape + ', AWSError>;\n';
+    code += this.generateDocString(operation.documentation, tabCount);
+    code += tabs(tabCount) + operationName + '(callback?: (err: AWSError, data: ' + outputShape + ') => void): Request<' + outputShape + ', AWSError>;\n';
+
+    return code;
+};
+
+/**
+ * Generates class method types for a waiter.
+ */
+TSGenerator.prototype.generateTypingsFromWaiters = function generateTypingsFromWaiters(className, waiterState, waiter, underlyingOperation, tabCount) {
+    var code = '';
+    tabCount = tabCount || 0;
+    var operationName = waiter.operation.charAt(0).toLowerCase() + waiter.operation.substring(1);
+
+    var docString = 'Waits for the ' + waiterState + ' state by periodically calling the underlying ' + className + '.' + operationName + 'operation every ' + waiter.delay + ' seconds (at most ' + waiter.maxAttempts + ' times).';
+    if (waiter.description) {
+        docString += ' ' + waiter.description;
+    }
+
+    // get input and output
+    var inputShape = '{}';
+    var outputShape = '{}';
+    if (underlyingOperation.input) {
+        inputShape = className + '.' + underlyingOperation.input.shape;
+    }
+    if (underlyingOperation.output) {
+        outputShape = className + '.' + underlyingOperation.output.shape;
+    }
+
+    code += this.generateDocString(docString, tabCount);
+    code += this.tabs(tabCount) + 'waitFor(state: "' + waiterState + '", params: ' + inputShape + ', callback?: (err: AWSError, data: ' + outputShape + ') => void): Request<' + outputShape + ', AWSError>;\n';
+    code += this.generateDocString(docString, tabCount);
+    code += this.tabs(tabCount) + 'waitFor(state: "' + waiterState + '", callback?: (err: AWSError, data: ' + outputShape + ') => void): Request<' + outputShape + ', AWSError>;\n';
```

Comment:
Same comment as above about required params. If the underlying operation has any required params, that perhaps we shouldn't generate this second typing where params is optional.